### PR TITLE
Fixed #33213 -- Doc'd testing code coverage in parallel and used it.

### PR DESCRIPTION
**Pull request by pauloxnet**
### Stats
| processes | tests | time     | db setup | db teardown | total run | cover |
| --------- | ----- | -------- | -------- | ----------- | --------- | ----- |
| 8         | 16190 | 192.076s | 26.254s  | 0.000s      | 223.999s  | 82%   |
| 1         | 16190 | 708.954s | 27.182s  | 0.000s      | 741.697s  | 82%   |

### CODE REVIEW

**Review by emarhin — APPROVED** [review]
that neat look

**Review by carltongibson — APPROVED** [review]
Hey @pauloxnet — thanks for this — (as far as I can tell :) it works well for me on macOS and Windows, so 👍 

I will ask @felixxm if he has comments.

**Comment by carltongibson** [comment]
Fix for https://code.djangoproject.com/ticket/33213?

File: docs/internals/contributing/writing-code/unit-tests.txt [hunk 1]
```diff
@@ -356,14 +356,19 @@ Contributors are encouraged to run coverage on the test suite to identify areas
...
 in :ref:`testing code coverage<topics-testing-code-coverage>`.
 
-Coverage should be run in a single process to obtain accurate statistics. To
-run coverage on the Django test suite using the standard test settings:
+To run coverage on the Django test suite using the standard test settings:
 
 .. console::
 
-   $ coverage run ./runtests.py --settings=test_sqlite --parallel=1
+   $ coverage run ./runtests.py --settings=test_sqlite
```
Comment (felixxm) [inline docs/internals/contributing/writing-code/unit-tests.txt:363]: Should we add `--concurrency=multiprocessing`? (see #15116) :thinking:

**Comment by felixxm** [comment]
@adamchainz Can you take a look? :eagle:

File: docs/internals/contributing/writing-code/unit-tests.txt (hunk 1 above)
Comment (pauloxnet) [reply docs/internals/contributing/writing-code/unit-tests.txt:363]: I've added `concurrency = multiprocessing` in [`.coveragerc`](https://github.com/django/django/pull/16435/files#diff-4717170cf23369601199611189b7024c618ee859d198d90334e270c4f1cbdf36R3)

Comment (felixxm) [reply docs/internals/contributing/writing-code/unit-tests.txt:363]: Right, good :+1:

File: tests/.coveragerc [hunk 2]
```diff
@@ -1,5 +1,7 @@
 [run]
 branch = True
+concurrency = multiprocessing
+data_file = .coverages/.coverage
```
Comment (adamchainz) [inline tests/.coveragerc:4]: why configure `data_file`? the default is normally sufficient and I can't see why not here

**Review by adamchainz — APPROVED** [review]
LGTM apart from the question about `data_file`

File: tests/.coveragerc (hunk 2 above)
Comment (pauloxnet) [reply tests/.coveragerc:4]: I customized the value of `data_file` for a better cleanup of the workspace when using `coverage` in parallel.

When used in parallel `coverage` creates a file for each process in the path where it runs.

These files are listed _(eg: in your editor)_ even if git-ignored until they are combined and sometimes you have to delete them by hand if something goes wrong.
```
.coverage.net.24830.761944
.coverage.net.24990.580886
.coverage.net.24991.642706
.coverage.net.24992.475125
.coverage.net.24993.937970
.coverage.net.24994.804635
.coverage.net.24995.747769
.coverage.net.24996.767985
```

Using the `.coverages/.coverage` value in `data_file` causes all intermediate files to be created inside the `.coverages` directory.

Now you will see _(eg: in your editor)_ only one git-ignored `.coverages` directory and if needed you can delete all old files left by deleting the whole directory in one go.
```bash
.coverages/
```

Comment (adamchainz) [reply tests/.coveragerc:4]: Okay fair enough!

**Comment by adamchainz** [comment]
> Do you have any other questions or concerns regarding this PR?

No, good work, thank you!

**Comment by felixxm** [comment]
I afraid we have to revert this change, as test coverage for async methods is no longer calculated.

**Comment by felixxm** [comment]
See #16655.

**Comment by pauloxnet** [comment]
> I afraid we have to revert this change, as test coverage for async methods is no longer calculated. See [a coverage report](https://djangoci.com/view/%C2%ADCoverage/job/django-coverage-postgresql/HTML_20Coverage_20Report/).

In this PR we only updated the documentation and one coverage setting.

@felixxm can you point me to the line in the CI configuration that executes tests?

**Comment by felixxm** [comment]
> @felixxm can you point me to the line in the CI configuration that executes tests?

Locally, I run tests as documented in this change:
```shell
$ coverage run runtests.py --settings=test_sqlite
$ coverage combine
$ coverage html
```
CI is doing exactly the same
```shell
coverage run runtests.py --settings=test_sqlite3
coverage combine
coverage html 
coverage xml
```
In both cases coverage for async code is missing or unreliable.

**Comment by pauloxnet** [comment]
> > @felixxm can you point me to the line in the CI configuration that executes tests?
> 
> Locally, I run tests as documented in this change:
> 
> ```shell
> $ coverage run runtests.py --settings=test_sqlite
> $ coverage combine
> $ coverage html
> ```
> 
> CI is doing exactly the same
> 
> ```shell
> coverage run runtests.py --settings=test_sqlite3
> coverage combine
> coverage html 
> coverage xml
> ```
> 
> In both cases coverage for async code is missing or unreliable.

I'm trying it locally to understand the motivation, and I'll write to you ASAP.

**Comment by pauloxnet** [comment]
> > @felixxm can you point me to the line in the CI configuration that executes tests?
> 
> Locally, I run tests as documented in this change:
> 
> ```shell
> $ coverage run runtests.py --settings=test_sqlite
> $ coverage combine
> $ coverage html
> ```
> 
> CI is doing exactly the same
> 
> ```shell
> coverage run runtests.py --settings=test_sqlite3
> coverage combine
> coverage html 
> coverage xml
> ```
> 
> In both cases coverage for async code is missing or unreliable.

I tried running the tests locally and found no problems.

If you think reverting the commit solves the problem, go ahead.

But in order to better investigate the problem I would ask you:
- from what in particular do you infer that tests for asynchronous methods are no longer computed?
- what coverage percentages did you expect and in which rows?

Over the next few days I would like to investigate this anomaly you reported and make sure I can run Django tests in parallel with global coverage.

**Comment by felixxm** [comment]
> I tried running the tests locally and found no problems.

Really? Do you have exactly the same numbers? :thinking: 

> from what in particular do you infer that tests for asynchronous methods are no longer computed?

Have you checked logs that I mentioned?

> what coverage percentages did you expect and in which rows?

For example, with this patch:
```
Coverage for /django/django/contrib/auth/middleware.py: 85%
```
image

Without this patch:
```
Coverage for /home/felixx/repo/django/django/contrib/auth/middleware.py: 91%
```
image

**Comment by pauloxnet** [comment]
>> what coverage percentages did you expect and in which rows?
> 
> For example, with this patch:
> 
> ```
> Coverage for /django/django/contrib/auth/middleware.py: 85%
> ```
> 
> Without this patch:
> 
> ```
> Coverage for /home/felixx/repo/django/django/contrib/auth/middleware.py: 91%
> ```

Ok, thanks. Actually locally I confirm that async methods are not covered. It's strange, but I'll try to analyze this behavior only next week, so go ahead with the revert PR.

**Pull request merged**